File: contrib/bin/xgbd/rf_evaluate.py
# Evaluate an xgboost descriptor model built with xgboost_make

import os
import sys

import joblib
import pandas as pd
import random_forest_model_pb2
from absl import app, flags, logging
from google.protobuf import text_format

from pytk.xgbd.linear_scaling_pb2 import LinearScalingData

FLAGS = flags.FLAGS

flags.DEFINE_string("mdir", "", "Model directory")

def read_rescaling_data(fname: str):
  """Read a linear_scaling_pb2.LinearScalingData serialized proto from `fname`.
  """
  with open(fname, "rb") as input:
    data = input.read()

  result = LinearScalingData()
  result.ParseFromString(data)

  return result

def get_model(mdir: str)->tuple:
  """Look for 'model_metadata.txt` in `mdir` and make sure it is OK
    Return a model instantiated from mdir/random_forest.joblib and the
    name of the response.
    If rescaling.dat is present, also return it.
  """
  failed = None, None, None
  fname = os.path.join(mdir, "model_metadata.txt")
  if not os.path.exists(fname):
    logging.error("%s not found", fname)
    return failed

  with open(fname, "r") as reader:
    text = reader.read()

  proto = text_format.Parse(text, random_forest_model_pb2.RandomForestModel())
  if not proto:
    logging.error("Cannot interpret as proto %s", text)
    return failed

  if not proto.response:
    logging.error("No response in %s", fname)
    return failed

  model_file = os.path.join(mdir, "random_forest.joblib")
  if not os.path.exists(model_file):
    logging.error("%s not found", model_file)
    return failed

  model = joblib.load(model_file)

  rescaling = os.path.join(mdir, 'rescaling.dat')
  if os.path.exists(rescaling):
    True
  else:
    rescaling = None

  return model, proto.response, rescaling

def random_forest_evaluate(mdir: str, fname: str)->bool:
  """Read `fname` as descriptors for a model in `mdir`
  """
  if not os.path.isdir(mdir):
    logging.error("Model directory %s not found", mdir)
    return False

  model, response, rescaling_fname = get_model(mdir)
  if not model:
    logging.error("Invalid mode in %s", mdir)
    return False

  if rescaling_fname is not None:
    rescaling = read_rescaling_data(rescaling_fname)
    # print(rescaling, file=sys.stderr)
  else:
    rescaling = None

  data = pd.read_csv(fname, sep=' ', header=0)

  logging.info("Evaluating %d rows", len(data))
  results = model.predict(data.iloc[:,1:])
  print(f"Id RF_{response}")
  if rescaling is None:
    for i in range(len(results)):
      print(f"{data.iloc[i,0]} {results[i]:.4f}")
  else:
    for i in range(len(results)):
      unscaled = (results[i] - rescaling.intercept) / rescaling.slope
      print(f"{data.iloc[i,0]} {unscaled:.4f}")

  return True

def main(argv):
  """Evaluate a random forest descriptor model.
  """
  if len(argv) == 1:
    logging.error("Must specify descriptor file as argument")
    return 1

  if not FLAGS.mdir:
    logging.error("must specify model directory via the --mdir option")
    return 1

  return random_forest_evaluate(FLAGS.mdir, argv[1])

if __name__ == '__main__':
  app.run(main)

File: contrib/bin/xgbd/rf_evaluate.rb
#!/usr/bin/env ruby

# Evaluate an RF model from either smiles or a descriptor file.

require 'set'
require 'google/protobuf'

c3tk_home = ENV['C3TK_HOME']
raise 'C3TK_HOME not defined' unless c3tk_home

require "#{c3tk_home}/bin/ruby/lib/iwcmdline"
require "#{c3tk_home}/bin/py/pytk/xgbd/random_forest_model_pb"

def usage
msg = <<-END
Scores a random_forest descriptor model built with rf_make.
Takes either a smiles file or a descriptor file,
If a smiles file is given, the descriptors required for the model will be computed by make_descriptors
  -mdir <dir>           model directory created by rf_make
  -smi                  input is a smiles file - model descriptors will be computed.
  -pipe                 use descriptor_pipeline for descriptor compuation,
                        only some descriptors are supported, but maybe faster.
  -j <threads>          send the -j option to make_descriptors.sh
  -v                    verbose output
END
  $stderr << msg

  exit(0)
end

# Return the proto mdir/model_metadata.dat
def get_model_metadata(mdir)
  fname = File.join(mdir, 'model_metadata.dat')

  unless File.size?(fname)
    $stderr << "Empty or missing model metadata #{fname}\n"
    return 1
  end

  return RandomForestModel::RandomForestModel.decode(File.read(fname))
end

def rf_evaluate_smiles(fname, mdir, proto, cl)
  descriptors = Set.new()
  proto.name_to_col.each do |name, col|
    descriptors.add(name.gsub(/_.*/, ""))
  end

  if cl.option_present('pipe')
    cmd = 'descriptor_pipeline.sh'
  else
    cmd = 'make_descriptors.sh'
  end

  if cl.option_present('j')
    j = cl.value('j')
    cmd << " -j #{j}"
  end

  tmpfile = File.join(ENV['TMPDIR'], "rf_evaluate_smiles_#{Process.uid}.#{Process.pid}.dat")

  descriptors.each do |d|
    cmd << " -#{d}"
  end
  cmd << " #{fname} > #{tmpfile}"
  $stderr << "Executing #{cmd}\n" if cl.option_present('v')

  system(cmd)
  unless File.size?(tmpfile)
    $stderr << "#{cmd} failed\n"
    return
  end

  rf_evaluate_descriptors(tmpfile, mdir, proto, cl)

  File.unlink(tmpfile)
end

def rf_evaluate_descriptors(fname, mdir, proto, cl)
  cmd = "rf_evaluate.sh -mdir #{mdir} #{fname}"

  $stderr << "Executing #{cmd}\n" if cl.option_present('v')
  system(cmd)
end

def main
  cl = IWCmdline.new("-v-mdir=dir-smi-pipe-j=ipos")

  if cl.unrecognised_options_encountered
    $stderr << "unrecognised_options_encountered\n"
    usage
  end

  unless cl.option_present('mdir')
    $stderr << "Must specify random forest model directory via the -mdir option\n"
    usage
  end

  if ARGV.empty?
    $stderr << "Insufficient arguments\n"
    usage
  end

  mdir = cl.value('mdir')

  proto = get_model_metadata(mdir)

  if cl.option_present('smi') || ARGV[0].match(/\.smi$/)
    rf_evaluate_smiles(ARGV[0], mdir, proto, cl)
  else
    rf_evaluate_descriptors(ARGV[0], mdir, proto, cl)
  end

end

main
